libs/core/langchain_core/runnables/graph.py

```diff
@@ -0,0 +1,133 @@
+        self.nodes.pop(node.id)
+        self.edges = [
+            edge
+            for edge in self.edges
+            if edge.source != node.id and edge.target != node.id
+        ]
+
+    def add_edge(self, source: Node, target: Node) -> Edge:
+        """Add an edge to the graph and return it."""
+        if source.id not in self.nodes:
+            raise ValueError(f"Source node {source.id} not in graph")
+        if target.id not in self.nodes:
+            raise ValueError(f"Target node {target.id} not in graph")
+        edge = Edge(source=source.id, target=target.id)
+        self.edges.append(edge)
+        return edge
+
+    def extend(self, graph: Graph) -> None:
+        """Add all nodes and edges from another graph.
+        Note this doesn't check for duplicates, nor does it connect the graphs."""
+        self.nodes.update(graph.nodes)
+        self.edges.extend(graph.edges)
+
+    def first_node(self) -> Optional[Node]:
+        """Find the single node that is not a target of any edge.
+        If there is no such node, or there are multiple, return None.
+        When drawing the graph this node would be the origin."""
+        targets = {edge.target for edge in self.edges}
+        found: List[Node] = []
+        for node in self.nodes.values():
+            if node.id not in targets:
+                found.append(node)
+        return found[0] if len(found) == 1 else None
+
+    def last_node(self) -> Optional[Node]:
+        """Find the single node that is not a source of any edge.
+        If there is no such node, or there are multiple, return None.
+        When drawing the graph this node would be the destination.
+        """
+        sources = {edge.source for edge in self.edges}
+        found: List[Node] = []
+        for node in self.nodes.values():
+            if node.id not in sources:
+                found.append(node)
+        return found[0] if len(found) == 1 else None
+
+    def trim_first_node(self) -> None:
+        """Remove the first node if it exists and has a single outgoing edge,
+        ie. if removing it would not leave the graph without a "first" node."""
+        first_node = self.first_node()
+        if first_node:
+            if (
+                len(self.nodes) == 1
+                or len([edge for edge in self.edges if edge.source == first_node.id])
+                == 1
+            ):
+                self.remove_node(first_node)
+
+    def trim_last_node(self) -> None:
+        """Remove the last node if it exists and has a single incoming edge,
+        ie. if removing it would not leave the graph without a "last" node."""
+        last_node = self.last_node()
+        if last_node:
+            if (
+                len(self.nodes) == 1
+                or len([edge for edge in self.edges if edge.target == last_node.id])
+                == 1
+            ):
+                self.remove_node(last_node)
+
+    def draw_ascii(self) -> str:
```

Comment:
should these be separate renderers and support different rendering types?  for example ascii, mermaid, graphviz, etc.

Reply:
yea that's why its a separate method